plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/domains/CastingDomain.kt

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.domains
+
+import org.jetbrains.kotlin.formver.embeddings.TypeEmbedding
+import org.jetbrains.kotlin.formver.viper.ast.*
+
+/**
+ * Casting domain to cast any type to any other type
+ *
+ * Viper Domain:
+ * ```
+ * domain Casting[A, B]  {
+ *
+ *   function cast(a: A, newType: Type): B
+ *
+ *   axiom null_cast {
+ *     (forall newType: Type :: { (cast((null_val(): Nullable[A]), NullableType(newType)): Nullable[B]) }
+ *       (cast((null_val(): Nullable[A]), NullableType(newType)): Nullable[B]) == (null_val(): Nullable[B]))
+ *   }
+ *
+ *   axiom type_of_cast {
+ *     (forall a: A, newType: Type :: { (typeOf((cast(a, newType): Nullable[B])): Type) }
+ *       (typeOf((cast(a, newType): Nullable[B])): Type) == newType)
+ *   }
+ * }
+ * ```
+ */
+object CastingDomain : BuiltinDomain("Casting") {
+    private val A = Type.TypeVar("A")
+    private val B = Type.TypeVar("B")
+    override val typeVars: List<Type.TypeVar> = listOf(A, B)
+
+    val a = Var("a", A)
+    val newType = Var("newType", TypeDomain.Type)
+
+    private val castFunc = createDomainFunc("cast", listOf(a.decl(), newType.decl()), B)
+
+    fun cast(exp: Exp, newType: Exp, newViperType: Type) =
+        funcApp(castFunc, listOf(exp, newType), mapOf(A to exp.type, B to newViperType))
+
+    // Prefer this cast method if you have access to a `TypeEmbedding`.
```

Comment:
What is the benefit of having a `TypeEmbedding` here?  Is it just so that we can have the Kotlin and Viper types packaged up?
It seems like a small thing to move all the domains over for.

Reply:
It's not just for having `TypeEmbedding`, the main reason is that `TypeDomain` relies on the `ClassEmbedding`s of the program (since it generates functions and axioms for the user classes) and all other domains besides `UnitDomain` depend on `TypeDomain`.